OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,506 @@
+				{
+					var inf = new Inflater(true);
+					var buffer = new byte[165535];
+					do
+					{
+						var bytesToExtract = cabFile.ReadUInt16();
+						RemainingArchiveStream -= 2u;
+						RemainingFileStream -= 2u;
+						inf.SetInput(GetBytes(bytesToExtract));
+						RemainingFileStream -= bytesToExtract;
+						while (!inf.IsNeedingInput)
+						{
+							var inflated = inf.Inflate(buffer);
+							dest.Write(buffer, 0, inflated);
+						}
+
+						inf.Reset();
+					}
+					while (RemainingFileStream > 0);
+				}
+				else
+				{
+					do
+					{
+						RemainingFileStream -= RemainingArchiveStream;
+						dest.Write(GetBytes(RemainingArchiveStream), 0, (int)RemainingArchiveStream);
+					}
+					while (RemainingFileStream > 0);
+				}
+			}
+
+			public byte[] GetBytes(uint count)
+			{
+				if (count < RemainingArchiveStream)
+				{
+					RemainingArchiveStream -= count;
+					return cabFile.ReadBytes((int)count);
+				}
+				else
+				{
+					var outArray = new byte[count];
+					var read = cabFile.Read(outArray, 0, (int)RemainingArchiveStream);
+					if (RemainingFileStream > RemainingArchiveStream)
+					{
+						NextFile();
+						RemainingArchiveStream -= (uint)cabFile.Read(outArray, read, (int)count - read);
+					}
+
+					return outArray;
+				}
+			}
+
+			public void Dispose()
+			{
+				cabFile.Dispose();
+			}
+
+			public void NextFile()
+			{
+				if (cabFile != null)
+					cabFile.Dispose();
+
+				++volumeNumber;
+				cabFile = GlobalFileSystem.Open("{0}{1}.cab".F(commonName, volumeNumber));
+				if (cabFile.ReadUInt32() != 0x28635349)
+					throw new InvalidDataException("Not an Installshield CAB package");
+				uint fileOffset;
+				if ((fileDes.Flags & FileSplit) != 0)
+				{
+					new CommonHeader(cabFile);
+					var head = new VolumeHeader(cabFile);
+					if (index == head.LastFileIndex)
+					{
+						if ((fileDes.Flags & FileCompressed) != 0)
+							RemainingArchiveStream = head.LastFileSizeCompressed;
+						else
+							RemainingArchiveStream = head.LastFileSizeExpanded;
+
+						fileOffset = head.LastFileOffset;
+					}
+					else if (index == head.FirstFileIndex)
+					{
+						if ((fileDes.Flags & FileCompressed) != 0)
+							RemainingArchiveStream = head.FirstFileSizeCompressed;
+						else
+							RemainingArchiveStream = head.FirstFileSizeExpanded;
+
+						fileOffset = head.FirstFileOffset;
+					}
+					else throw new Exception("Cannot Resolve Remaining Stream");
+				}
+				else
+				{
+					if ((fileDes.Flags & FileCompressed) != 0)
+						RemainingArchiveStream = fileDes.CompressedSize;
+					else
+						RemainingArchiveStream = fileDes.ExpandedSize;
+
+					fileOffset = fileDes.DataOffset;
+				}
+
+				cabFile.Seek(fileOffset, SeekOrigin.Begin);
+			}
+		}
+#endregion
+
+		readonly Stream hdrFile;
+		CommonHeader commonHeader;
+		CabDescriptor cabDescriptor;
+		List<uint> directoryTable;
+		Dictionary<uint, string> directoryNames = new Dictionary<uint, string>();
```

Comment:
These dictionaries (and maybe some of the other members?) can be readonly (note: this doesn't imply that the dictionary itself is readonly, just the reference to it).